arduino-ide-extension/src/node/sketches-service-impl.ts

```diff
@@ -510,26 +517,104 @@ export class SketchesServiceImpl
+      try {
+        const otherInoBasename = `${destinationFolderBasename}.ino`;
+        const otherInoPath = join(source, otherInoBasename);
+        const stat = await fs.stat(otherInoPath);
+        if (stat.isFile()) {
+          const message = nls.localize(
+            'arduino/sketch/sketchAlreadyContainsThisFileError',
+            "The sketch already contains a file named '{0}'",
+            otherInoBasename
+          );
+          // if can read the file, it will be a collision
+          throw SketchesError.SketchAlreadyContainsThisFile(
+            message,
+            sourceFolderBasename,
+            destinationFolderBasename,
+            otherInoBasename
+          );
+        }
+        // Otherwise, it's OK, if it is an existing directory
+      } catch (err) {
+        // It's OK if the file is missing.
+        if (!ErrnoException.isENOENT(err)) {
+          throw err;
+        }
+      }
+    }
+
+    let filter: CopyOptions['filter'];
     if (onlySketchFiles) {
-      const sketchFilePaths = Sketch.uris(sketch).map(FileUri.fsPath);
-      filter = (file: { path: string }) => sketchFilePaths.includes(file.path);
+      // The Windows paths, can be a trash (see below). Hence, it must be resolved with Node.js.
+      // After resolving the path, the drive letter is still a gamble (can be upper or lower case) and could result in a false negative match.
+      // Here, all sketch file paths must be resolved by Node.js, to provide the same drive letter casing.
+      const sketchFilePaths = await Promise.all(
+        Sketch.uris(sketch)
+          .map(FileUri.fsPath)
+          .map((path) => fs.realpath(path))
+      );
+      filter = async (s) => {
+        // On Windows, the source path could start with a complete trash. For example, \\\\?\\c:\\Users\\kittaakos\\AppData\\Local\\Temp\\.arduinoIDE-unsaved20231024-9300-1hp64fi.g8yh\\sketch_nov24d.
+        // The path must be resolved.
+        const resolvedSource = await fs.realpath(s);
+        if (sketchFilePaths.includes(resolvedSource)) {
+          return true;
+        }
+        const stat = await fs.stat(resolvedSource);
+        if (stat.isFile()) {
+          return false;
+        }
+        // Copy the folder if any of the sketch file path starts with this folder
+        return sketchFilePaths.some((sketchFilePath) =>
+          sketchFilePath.startsWith(resolvedSource)
+        );
+      };
     } else {
       filter = () => true;
     }
-    const cpyModule = await import('cpy');
-    const cpy = cpyModule.default;
-    await cpy(sourceFolderBasename, destination, {
-      rename: (basename) =>
-        sourceFolderBasename !== destinationFolderBasename &&
-        basename === `${sourceFolderBasename}.ino`
-          ? `${destinationFolderBasename}.ino`
-          : basename,
+
+    const tempRoot = await this.createTempFolder();
+    const temp = join(tempRoot, destinationFolderBasename);
+    await fs.mkdir(temp, { recursive: true });
```

Comment:
Did you consider using `mkdtemp` here instead of `mkdir`? It don't believe it really matters, I bring it up just because it may be a more natural choice.

Reply:
+1 for using Node.js, but it does not work in this scenario. The sketch folder naming must follow specific rules (see the [spec](https://arduino.github.io/arduino-cli/dev/sketch-specification/#sketch-folders-and-files)). [`mkdtemp`](https://nodejs.org/api/fs.html#fspromisesmkdtempprefix-options) creates a folder with random characters. That's not what the tests want:


> A unique directory name is generated by appending six random characters to the end of the provided `prefix`.


```js
async function run() {
    const path = require('node:path');
    const { promises: fs } = require('node:fs');
    const os = require('node:os');
    const tempPath = await fs.mkdtemp(path.join(os.tmpdir(), 'foo'));
    console.log(tempPath);
}

run();
```

Output:
```
/var/folders/z1/xkw1yh5n7rz4n8djprp1mdn80000gn/T/fooMVXPWZ
```